File: agent-packs/eval-pilot/engine-ts/src/loaders/markdown.ts
/**
 * Parse a Markdown `*.eval.md` file into an {@link EvalSpec}.
 *
 * A spec reads top-to-bottom as arrange → act → assert. Prose (prompts, judge
 * criteria) is plain text; structured knobs live in small fenced `yaml` blocks.
 */

import { readFileSync } from "node:fs";
import * as path from "node:path";
import yaml from "js-yaml";
import {
  makeActSpec,
  makeAssertionSpec,
  makeEvalSpec,
  makeFileCopy,
  makeJudgeSpec,
  makeMetricSpec,
  makeSetupSpec,
  makeStageSpec,
  type ActSpec,
  type AssertionSpec,
  type EvalKind,
  type EvalSpec,
  type FileCopy,
  type JudgeSpec,
  type MetricSpec,
  type SetupSpec,
} from "../spec.js";

const FRONTMATTER = /^\ufeff?---\s*\n([\s\S]*?)\n---\s*\n/;
const FENCE = /```([^\n`]*)\n([\s\S]*?)```/g;
const HEADING = /^(#{1,6})\s+(.*)$/gm;

const LIST_KINDS = new Set([
  "contains",
  "not_contains",
  "prose_contains",
  "stdout_contains",
  "matches",
  "glob_count",
  "json_path",
  "json_empty",
  "section_contains",
  "section_not_contains",
  "file_exists",
  "file_absent",
]);

/** Raised when a `*.eval.md` file cannot be parsed into a spec. */
export class EvalParseError extends Error {}

/** Load and parse a `*.eval.md` file from disk. */
export function loadMarkdownEval(specPath: string): EvalSpec {
  const text = readFileSync(specPath, "utf-8");
  const spec = parseMarkdownEval(text, specPath);
  spec.base_dir = path.dirname(specPath);
  return spec;
}

/** Parse `*.eval.md` source text into an {@link EvalSpec}. */
export function parseMarkdownEval(text: string, specPath?: string): EvalSpec {
  const [front, body] = splitFrontmatter(text);
  const { title, summary: blockquote, sections } = splitSections(body);

  const name = front.name || slug(title) || "unnamed-eval";
  const kind = String(front.kind ?? "none").toLowerCase() as EvalKind;
  const summary = String(front.summary ?? blockquote ?? "").trim();

  const descSection = sections.get("description");
  let description = "";
  if (descSection !== undefined) description = descSection.prose().trim();
  if (!description) description = String(front.description ?? "").trim();

  const setup = parseSetup(sections.get("setup"));
  const act = parseAct(sections.get("act"), front);
  const { assertions, judges, metrics } = parseAssert(sections.get("assert"));

  return makeEvalSpec({
    name,
    target: front.target ?? null,
    kind,
    tags: asList(front.tags).map(String),
    timeout: Number(front.timeout ?? 600),
    summary,
    description: description || summary,
    setup,
    act,
    assertions,
    judges,
    metrics,
    spec_path: specPath ?? null,
    base_dir: specPath ? path.dirname(specPath) : null,
  });
}

// ---- structural parsing -------------------------------------------------

function splitFrontmatter(text: string): [Record<string, any>, string] {
  const m = FRONTMATTER.exec(text);
  if (!m) return [{}, text];
  let data: unknown;
  try {
    data = yaml.load(m[1]!) ?? {};
  } catch (exc) {
    throw new EvalParseError(`invalid frontmatter YAML: ${(exc as Error).message}`);
  }
  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    throw new EvalParseError("frontmatter must be a YAML mapping");
  }
  return [data as Record<string, any>, text.slice(m.index + m[0].length)];
}

class Section {
  raw: string;
  blocks: Array<[string, string]>;

  constructor(raw: string) {
    this.raw = raw;
    this.blocks = [];
    for (const m of raw.matchAll(new RegExp(FENCE.source, "g"))) {
      this.blocks.push([m[1]!.trim().toLowerCase(), m[2]!]);
    }
  }

  block(lang: string): string | null {
    for (const [blang, content] of this.blocks) {
      if (blang === lang) return content;
    }
    return null;
  }

  /** Section text with fenced blocks and blockquotes stripped. */
  prose(): string {
    const stripped = this.raw.replace(new RegExp(FENCE.source, "g"), "");
    const lines = stripped
      .split("\n")
      .filter((ln) => !ln.trim().startsWith(">"));
    return lines.join("\n").trim();
  }
}

interface SplitResult {
  title: string;
  summary: string;
  sections: Map<string, Section>;
}

function splitSections(body: string): SplitResult {
  let title = "";
  let summary = "";
  const sections = new Map<string, Section>();

  const masked = maskFences(body);
  const matches = [...masked.matchAll(new RegExp(HEADING.source, "gm"))];
  const firstStart = matches.length ? matches[0]!.index! : body.length;
  const preamble = body.slice(0, firstStart);

  for (const h of matches) {
    const level = h[1]!.length;
    const headingText = h[2]!.trim();
    if (level === 1 && !title) title = headingText;
  }
  if (!title) {
    const m = /^#\s+(.*)$/m.exec(preamble);
    if (m) title = m[1]!.trim();
  }

  // Collect level-2 sections (## Setup / ## Act / ## Assert ...).
  const h2 = matches.filter((h) => h[1]!.length === 2);
  for (let i = 0; i < h2.length; i++) {
    const h = h2[i]!;
    const start = h.index! + h[0].length;
    const end = i + 1 < h2.length ? h2[i + 1]!.index! : body.length;
    const nm = h[2]!.trim().toLowerCase();
    sections.set(nm, new Section(body.slice(start, end)));
  }

  // The `> summary` blockquote lives above the first ## section.
  const headerEnd = h2.length ? h2[0]!.index! : body.length;
  const qm = /^>\s?(.*)$/m.exec(body.slice(0, headerEnd));
  if (qm) summary = qm[1]!.trim();

  return { title, summary, sections };
}

/** Blank fenced regions while preserving length + newline positions. */
function maskFences(text: string): string {
  return text.replace(new RegExp(FENCE.source, "g"), (block) =>
    [...block].map((ch) => (ch === "\n" ? "\n" : " ")).join(""),
  );
}

// ---- section builders ---------------------------------------------------

function parseSetup(section: Section | undefined): SetupSpec {
  if (section === undefined) return makeSetupSpec();
  const raw = section.block("yaml") ?? section.block("json");
  let data = raw ? parseYaml(raw) : {};
  if (data === null || data === undefined) data = {};
  if (typeof data !== "object" || Array.isArray(data)) {
    throw new EvalParseError("## Setup yaml block must be a mapping");
  }
  const d = data as Record<string, any>;
  const stageData = d.stage ?? {};
  const stage = makeStageSpec({
    agent: stageData.agent ?? null,
    skill: stageData.skill ?? null,
    all: Boolean(stageData.all ?? false),
    include_skills: Boolean(stageData.include_skills ?? true),
  });
  const files: FileCopy[] = [];
  for (const f of asList(d.files)) {
    if (typeof f === "string") {
      files.push(makeFileCopy(f));
    } else if (f && typeof f === "object") {
      files.push(makeFileCopy(String(f.copy), String(f.dest ?? ".")));
    }
  }
  return makeSetupSpec({ stage, files });
}

function parseAct(
  section: Section | undefined,
  front: Record<string, any>,
): ActSpec | null {
  let prompt: string | null = null;
  let agent = front.agent ?? null;
  let skill = front.skill ?? null;
  let timeout = front.timeout ?? null;

  if (section !== undefined) {
    prompt = section.block("prompt");
    if (prompt === null) {
      const raw = section.block("yaml");
      if (raw) {
        const data = (parseYaml(raw) ?? {}) as Record<string, any>;
        prompt = data.prompt ?? null;
        agent = data.agent ?? agent;
        skill = data.skill ?? skill;
        timeout = data.timeout ?? timeout;
      }
    }
    if (prompt === null) prompt = section.prose();
  }
  if (!prompt) return null;
  const normalizedPrompt = prompt.endsWith("\n")
    ? prompt
    : prompt.replace(/\n+$/, "") + "\n";
  return makeActSpec({
    prompt: normalizedPrompt,
    agent,
    skill,
    timeout: timeout !== null && timeout !== undefined ? Number(timeout) : null,
  });
}

function parseAssert(section: Section | undefined): {
  assertions: AssertionSpec[];
  judges: JudgeSpec[];
  metrics: MetricSpec[];
} {
  const assertions: AssertionSpec[] = [];
  const judges: JudgeSpec[] = [];
  const metrics: MetricSpec[] = [];
  if (section === undefined) return { assertions, judges, metrics };

  const raw = section.block("yaml") ?? section.block("json");
  let data = raw ? parseYaml(raw) : {};
  if (data && (typeof data !== "object" || Array.isArray(data))) {
    throw new EvalParseError("## Assert yaml block must be a mapping");
  }
  const d = (data ?? {}) as Record<string, any>;

  // files: {exists: [...], absent: [...]}
  const files = d.files ?? {};
  if (files.exists) {
    assertions.push(
      makeAssertionSpec({
        kind: "file_exists",
        args: { paths: asList(files.exists) },
      }),
    );
  }
  if (files.absent) {
    assertions.push(
      makeAssertionSpec({
        kind: "file_absent",
        args: { paths: asList(files.absent) },
      }),
    );
  }

  // direct list kinds
  for (const kind of LIST_KINDS) {
    for (const entry of asList(d[kind])) {
      assertions.push(assertionFromEntry(kind, entry));
    }
  }

  // generic escape hatch
  for (const entry of asList(d.asserts)) {
    if (!entry || typeof entry !== "object" || !("kind" in entry)) {
      throw new EvalParseError("each 'asserts' entry needs a 'kind'");
    }
    const e = { ...(entry as Record<string, any>) };
    const kind = e.kind;
    const nm = e.name ?? null;
    delete e.kind;
    delete e.name;
    assertions.push(makeAssertionSpec({ kind, args: e, name: nm }));
  }

  // judges (single mapping or list)
  const judgeBlock = d.judge;
  if (judgeBlock !== undefined && judgeBlock !== null) {
    const entries = Array.isArray(judgeBlock) ? judgeBlock : [judgeBlock];
    for (const e of entries) judges.push(judgeFromEntry(e));
  }
  // standalone ```judge fenced block => criteria only
  const jc = section.block("judge");
  if (jc) judges.push(makeJudgeSpec({ criteria: jc.trim() }));

  for (const e of asList(d.metrics)) {
    metrics.push(metricFromEntry(e));
  }

  return { assertions, judges, metrics };
}

function assertionFromEntry(kind: string, entry: unknown): AssertionSpec {
  if (typeof entry === "string") {
    return makeAssertionSpec({ kind, args: { text: entry } });
  }
  if (entry && typeof entry === "object") {
    const e = { ...(entry as Record<string, any>) };
    const nm = e.name ?? null;
    delete e.name;
    return makeAssertionSpec({ kind, args: e, name: nm });
  }
  throw new EvalParseError(`invalid ${kind} entry: ${JSON.stringify(entry)}`);
}

function judgeFromEntry(e: any): JudgeSpec {
  if (!e || typeof e !== "object" || !("criteria" in e)) {
    throw new EvalParseError("each judge needs 'criteria'");
  }
  return makeJudgeSpec({
    criteria: String(e.criteria),
    artifact: e.artifact ?? null,
    threshold: Number(e.threshold ?? 0.7),
    name: String(e.name ?? "judge"),
    golden: asList(e.golden).map(String),
  });
}

function metricFromEntry(e: any): MetricSpec {
  if (!e || typeof e !== "object" || !("name" in e)) {
    throw new EvalParseError("each metric needs 'name'");
  }
  const baseline = e.baseline ?? "last";
  let baselineValue = e.baseline_value ?? null;
  const strategy = typeof baseline === "string" ? baseline : "pinned";
  if (typeof baseline === "number") baselineValue = Number(baseline);
  return makeMetricSpec({
    name: String(e.name),
    value: e.value,
    direction: String(e.direction ?? "higher_is_better"),
    unit: String(e.unit ?? ""),
    baseline_strategy: strategy,
    baseline: baselineValue,
    window: Number(e.window ?? 5),
    tolerance: Number(e.tolerance ?? 0.0),
    tolerance_pct: Number(e.tolerance_pct ?? 0.0),
    gate: Boolean(e.gate ?? false),
  });
}

// ---- small utilities ----------------------------------------------------

function parseYaml(raw: string): unknown {
  try {
    return yaml.load(raw);
  } catch (exc) {
    throw new EvalParseError(`invalid YAML block: ${(exc as Error).message}`);
  }
}

function asList(value: unknown): any[] {
  if (value === null || value === undefined) return [];
  if (Array.isArray(value)) return [...value];
  return [value];
}

function slug(text: string): string {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

File: agent-packs/eval-pilot/engine/src/evalpilot/loaders/markdown.py
"""Parse a Markdown ``*.eval.md`` file into an :class:`~evalpilot.spec.EvalSpec`.

The Markdown surface is the default because most of an eval *is* prose —
prompts and judge criteria read naturally as text — while the few structured
knobs (staging, file assertions, metrics) live in small fenced ``yaml``
blocks. A spec reads top-to-bottom as arrange → act → assert.

File shape::

    ---
    name: my-eval
    target: my-agent
    kind: agent            # agent | skill | none
    tags: [smoke, judge]
    timeout: 900
    ---

    # Human title
    > One-line summary (compact view).

    ## Description
    Multi-paragraph, human-readable explanation of what this eval checks and
    why. Shown in detailed output and used as the authoring intent when a
    user writes the description first and asks an agent to implement the rest.

    ## Setup
    ```yaml
    stage: { agent: my-agent }
    files: [{ copy: "fixtures/**", dest: "." }]
    ```

    ## Act
    ```prompt
    ...the prompt sent to the SUT...
    ```

    ## Assert
    ```yaml
    files: { exists: [out.md], absent: ["**/SECRET"] }
    contains: [{ path: out.md, text: "Status: draft" }]
    judge:   { artifact: out.md, threshold: 0.7, criteria: "..." }
    metrics: [{ name: judge_score, value: $judge.score,
                direction: higher_is_better, baseline: rolling_mean,
                tolerance: 0.1 }]
    ```

Any ``## Assert`` list key maps to an assertion kind; ``asserts:`` is the
generic escape hatch (``{kind: ..., ...}``).
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Optional

import yaml

from ..spec import (
    ActSpec,
    AssertionSpec,
    EvalSpec,
    FileCopy,
    JudgeSpec,
    MetricSpec,
    SetupSpec,
    StageSpec,
)

_FRONTMATTER = re.compile(r"^\ufeff?---\s*\n(.*?)\n---\s*\n", re.DOTALL)
_FENCE = re.compile(r"```([^\n`]*)\n(.*?)```", re.DOTALL)
_HEADING = re.compile(r"^(#{1,6})\s+(.*)$", re.MULTILINE)

# Assert-block list keys that map directly onto an assertion kind.
_LIST_KINDS = {
    "contains", "not_contains", "prose_contains", "stdout_contains",
    "matches", "glob_count", "json_path", "json_empty",
    "section_contains", "section_not_contains", "file_exists", "file_absent",
}


class EvalParseError(ValueError):
    """Raised when a ``*.eval.md`` file cannot be parsed into a spec."""


def load_markdown_eval(path: Path) -> EvalSpec:
    """Load and parse a ``*.eval.md`` file from disk."""
    path = Path(path)
    text = path.read_text(encoding="utf-8")
    spec = parse_markdown_eval(text, spec_path=path)
    spec.base_dir = path.parent
    return spec


def parse_markdown_eval(text: str, *, spec_path: Optional[Path] = None) -> EvalSpec:
    """Parse ``*.eval.md`` source text into an :class:`EvalSpec`."""
    front, body = _split_frontmatter(text)
    title, blockquote, sections = _split_sections(body)

    name = front.get("name") or _slug(title) or "unnamed-eval"
    kind = str(front.get("kind", "none")).lower()
    summary = str(front.get("summary") or blockquote or "").strip()
    desc_section = sections.get("description")
    description = ""
    if desc_section is not None:
        description = desc_section.prose().strip()
    if not description:
        description = str(front.get("description") or "").strip()
    setup = _parse_setup(sections.get("setup"))
    act = _parse_act(sections.get("act"), front)
    assertions, judges, metrics = _parse_assert(sections.get("assert"))

    spec = EvalSpec(
        name=name,
        target=front.get("target"),
        kind=kind,
        tags=_as_list(front.get("tags")),
        timeout=float(front.get("timeout", 600)),
        summary=summary,
        description=description or summary,
        setup=setup,
        act=act,
        assertions=assertions,
        judges=judges,
        metrics=metrics,
        spec_path=str(spec_path) if spec_path else None,
        base_dir=spec_path.parent if spec_path else None,
    )
    return spec


# ---- structural parsing -------------------------------------------------


def _split_frontmatter(text: str) -> tuple[dict, str]:
    m = _FRONTMATTER.match(text)
    if not m:
        return {}, text
    try:
        data = yaml.safe_load(m.group(1)) or {}
    except yaml.YAMLError as exc:
        raise EvalParseError(f"invalid frontmatter YAML: {exc}") from exc
    if not isinstance(data, dict):
        raise EvalParseError("frontmatter must be a YAML mapping")
    return data, text[m.end():]


class _Section:
    def __init__(self, raw: str) -> None:
        self.raw = raw
        self.blocks: list[tuple[str, str]] = [
            (lang.strip().lower(), content)
            for lang, content in _FENCE.findall(raw)
        ]

    def block(self, lang: str) -> Optional[str]:
        for blang, content in self.blocks:
            if blang == lang:
                return content
        return None

    def prose(self) -> str:
        """Section text with fenced blocks and blockquotes stripped."""
        stripped = _FENCE.sub("", self.raw)
        lines = [ln for ln in stripped.splitlines() if not ln.strip().startswith(">")]
        return "\n".join(lines).strip()


def _split_sections(body: str) -> tuple[str, str, dict[str, _Section]]:
    """Return (title, summary, {section_name_lower: _Section}).

    The second value is the one-line ``> blockquote`` summary that lives under
    the ``# Title`` (above the first ``## Section``).
    """
    title = ""
    summary = ""
    sections: dict[str, _Section] = {}

    # Blank out fenced regions so headings *inside* code don't split sections.
    masked = _mask_fences(body)

    matches = list(_HEADING.finditer(masked))
    first_start = matches[0].start() if matches else len(body)
    preamble = body[:first_start]

    for h in matches:
        level = len(h.group(1))
        heading_text = h.group(2).strip()
        if level == 1 and not title:
            title = heading_text
    if not title:
        m = re.search(r"^#\s+(.*)$", preamble, re.MULTILINE)
        if m:
            title = m.group(1).strip()

    # Collect level-2 sections (## Setup / ## Act / ## Assert ...).
    h2 = [h for h in matches if len(h.group(1)) == 2]
    for i, h in enumerate(h2):
        start = h.end()
        end = h2[i + 1].start() if i + 1 < len(h2) else len(body)
        name = h.group(2).strip().lower()
        sections[name] = _Section(body[start:end])

    # The `> summary` blockquote lives above the first ## section (usually
    # right under the # Title). Scan that whole header region.
    header_end = h2[0].start() if h2 else len(body)
    qm = re.search(r"^>\s?(.*)$", body[:header_end], re.MULTILINE)
    if qm:
        summary = qm.group(1).strip()

    return title, summary, sections


def _mask_fences(text: str) -> str:
    """Blank fenced regions while preserving length + newline positions.

    Heading detection runs on the masked text so ``#`` lines *inside* code
    fences don't split sections; offsets must stay aligned with the original
    body, so each fence is replaced char-for-char (newlines kept, others → space).
    """

    def _blank(m: "re.Match") -> str:
        return "".join("\n" if ch == "\n" else " " for ch in m.group(0))

    return _FENCE.sub(_blank, text)


# ---- section builders ---------------------------------------------------


def _parse_setup(section: Optional[_Section]) -> SetupSpec:
    if section is None:
        return SetupSpec()
    raw = section.block("yaml") or section.block("json")
    data = _yaml(raw) if raw else {}
    if data is None:
        data = {}
    if not isinstance(data, dict):
        raise EvalParseError("## Setup yaml block must be a mapping")
    stage_data = data.get("stage") or {}
    stage = StageSpec(
        agent=stage_data.get("agent"),
        skill=stage_data.get("skill"),
        all=bool(stage_data.get("all", False)),
        include_skills=bool(stage_data.get("include_skills", True)),
    )
    files = []
    for f in data.get("files", []) or []:
        if isinstance(f, str):
            files.append(FileCopy(copy=f))
        elif isinstance(f, dict):
            files.append(FileCopy(copy=str(f["copy"]), dest=str(f.get("dest", "."))))
    return SetupSpec(stage=stage, files=files)


def _parse_act(section: Optional[_Section], front: dict) -> Optional[ActSpec]:
    prompt = None
    agent = front.get("agent")
    skill = front.get("skill")
    timeout = front.get("timeout")
    if section is not None:
        prompt = section.block("prompt")
        if prompt is None:
            # A yaml block may carry {prompt, agent, skill, timeout}.
            raw = section.block("yaml")
            if raw:
                data = _yaml(raw) or {}
                prompt = data.get("prompt")
                agent = data.get("agent", agent)
                skill = data.get("skill", skill)
                timeout = data.get("timeout", timeout)
        if prompt is None:
            prompt = section.prose()
    if not prompt:
        return None
    return ActSpec(
        prompt=prompt.rstrip("\n") + "\n" if not prompt.endswith("\n") else prompt,
        agent=agent,
        skill=skill,
        timeout=float(timeout) if timeout is not None else None,
    )


def _parse_assert(
    section: Optional[_Section],
) -> tuple[list[AssertionSpec], list[JudgeSpec], list[MetricSpec]]:
    assertions: list[AssertionSpec] = []
    judges: list[JudgeSpec] = []
    metrics: list[MetricSpec] = []
    if section is None:
        return assertions, judges, metrics

    raw = section.block("yaml") or section.block("json")
    data = _yaml(raw) if raw else {}
    if data and not isinstance(data, dict):
        raise EvalParseError("## Assert yaml block must be a mapping")
    data = data or {}

    # files: {exists: [...], absent: [...]}
    files = data.get("files") or {}
    if files.get("exists"):
        assertions.append(AssertionSpec(kind="file_exists",
                                        args={"paths": _as_list(files["exists"])}))
    if files.get("absent"):
        assertions.append(AssertionSpec(kind="file_absent",
                                        args={"paths": _as_list(files["absent"])}))

    # direct list kinds
    for kind in _LIST_KINDS:
        for entry in data.get(kind, []) or []:
            assertions.append(_assertion_from_entry(kind, entry))

    # generic escape hatch
    for entry in data.get("asserts", []) or []:
        if not isinstance(entry, dict) or "kind" not in entry:
            raise EvalParseError("each 'asserts' entry needs a 'kind'")
        e = dict(entry)
        kind = e.pop("kind")
        name = e.pop("name", None)
        assertions.append(AssertionSpec(kind=kind, args=e, name=name))

    # judges (single mapping or list)
    judge_block = data.get("judge")
    if judge_block is not None:
        entries = judge_block if isinstance(judge_block, list) else [judge_block]
        for e in entries:
            judges.append(_judge_from_entry(e))
    # standalone ```judge fenced block => criteria only
    jc = section.block("judge")
    if jc:
        judges.append(JudgeSpec(criteria=jc.strip()))

    for e in data.get("metrics", []) or []:
        metrics.append(_metric_from_entry(e))

    return assertions, judges, metrics


def _assertion_from_entry(kind: str, entry) -> AssertionSpec:
    if isinstance(entry, str):
        return AssertionSpec(kind=kind, args={"text": entry})
    if isinstance(entry, dict):
        e = dict(entry)
        name = e.pop("name", None)
        return AssertionSpec(kind=kind, args=e, name=name)
    raise EvalParseError(f"invalid {kind} entry: {entry!r}")


def _judge_from_entry(e: dict) -> JudgeSpec:
    if not isinstance(e, dict) or "criteria" not in e:
        raise EvalParseError("each judge needs 'criteria'")
    return JudgeSpec(
        criteria=str(e["criteria"]),
        artifact=e.get("artifact"),
        threshold=float(e.get("threshold", 0.7)),
        name=str(e.get("name", "judge")),
        golden=_as_list(e.get("golden")),
    )


def _metric_from_entry(e: dict) -> MetricSpec:
    if not isinstance(e, dict) or "name" not in e:
        raise EvalParseError("each metric needs 'name'")
    baseline = e.get("baseline", "last")
    baseline_value = e.get("baseline_value")
    strategy = str(baseline) if isinstance(baseline, str) else "pinned"
    if isinstance(baseline, (int, float)):
        baseline_value = float(baseline)
    return MetricSpec(
        name=str(e["name"]),
        value=e.get("value"),
        direction=str(e.get("direction", "higher_is_better")),
        unit=str(e.get("unit", "")),
        baseline_strategy=strategy,
        baseline=baseline_value,
        window=int(e.get("window", 5)),
        tolerance=float(e.get("tolerance", 0.0)),
        tolerance_pct=float(e.get("tolerance_pct", 0.0)),
        gate=bool(e.get("gate", False)),
    )


# ---- small utilities ----------------------------------------------------


def _yaml(raw: str):
    try:
        return yaml.safe_load(raw)
    except yaml.YAMLError as exc:
        raise EvalParseError(f"invalid YAML block: {exc}") from exc


def _as_list(value) -> list:
    if value is None:
        return []
    if isinstance(value, (list, tuple)):
        return list(value)
    return [value]


def _slug(text: str) -> str:
    return re.sub(r"[^a-z0-9]+", "-", text.lower()).strip("-")


__all__ = ["load_markdown_eval", "parse_markdown_eval", "EvalParseError"]
